### DISCUSSION TITLE
selectFromResult should only re-invoke the selector if it has changed

### DISCUSSION THREAD

**Discussion by kkirby** [reduxjs/redux-toolkit#3975] [Ideas]
For example, when using `selectFromResult` to select a specific post via ID from an array of posts, the selector function should not be executed if the post ID did not change.

Basically, memoize the result unless the function has changed, in which case, re-run the function.

```javascript
function MyComponent({ postID }) {
  const selector = React.useCallback(({data}) => {
    return {
      data: data != null ? data.items.find(x => x.id === postID) : null,
    };
  }, [postID]);
  const { data, error, isLoading } = useGetExampleDataQuery(undefined, {
    selectFromResult: selector,
  });
```

This may be completely against the intended usage of selectFromResult, but I have seen code in the wild that uses `selectFromResult` this way.

**Comment by EskiMojo14** [comment]
what if the post being selected changed? how would you know without re-invoking the selector?

**Reply by kkirby** [reply]
If the data it depends on (the response data) changes, then of course the function should be called again.

**Comment by phryneas** [comment]
I might be missing something, but that's pretty much how it works?

Either upstream data (or loading states, other things that `selectFromResult` can work on) changes, or the `selectFromResult` function itself changes - it should only re-execute `selectFromResult` in one of those two cases.

**Reply by phryneas** [reply]
https://github.com/reduxjs/redux-toolkit/blob/72206a26551a5b756c65d173a98bc6a95694b24a/packages/toolkit/src/query/react/buildHooks.ts#L926-L931

Here you see the code - it will be created as a memoized selector. It shouldn't get re-executed unless data or `selectFromResult` itself changes.

**Reply by kkirby** [reply]
That memoizes the function itself, not the result. The function will still be invoked on every render.

**Reply by kkirby** [reply]
Here's a code sandbox that shows what I mean: https://codesandbox.io/p/sandbox/focused-tdd-j23jzv

When you click on `Show index 0` and `Show index 1`, it changes which index is being selected inside `selectFromResult`. If you click one button 3+ times, it will call `selectFromResult` each time the component renders. The component re-renders on every button click because it's always being passed new children containing the total button presses.

**Reply by phryneas** [reply]
That *will* keep `selectFromResult` from being invoked on every render. We might have a miscommunication going on here - do you mean something else than that?

**Reply by markerikson** [reply]
The child component is _already_ going to re-render, because you called `setState()` in the parent component and passed in a new object reference (which forces `<App>` to re-render), and React re-renders recursively by default:

- https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/#standard-render-behavior

`selectFromResult` isn't going to have an effect on whether `MyComponent` renders in that situation, because it's _already_ rendering.

(I may be missing something about this discussion, though. It's late here and I'm juggling tasks.)

**Reply by kkirby** [reply]
Yes, however, `selectFromResult` doesn't need to be invoked again on every render. If the function itself has not changed, there is no reason to think the return value of it would have changed either. My thought is that the result of the selector function should be cached and should only be invalidated if a new function is provided (or if the data being sent to the selector has changed).

In my example in my first post, the function I'm passing to `selectFromResult` will only change if `postID` changes.

Hopefully that makes sense?

**Reply by phryneas** [reply]
It makes sense, and it is exactly what the code that I linked to *should* do - but I see in your reproduction that it doesn't, which is deeply confusing.

**Reply by phryneas** [reply]
Although the way we call that might be at fault here, with `querySelector(state, lastValue.current)`, where `lastValue.current` could change a tad more often. 🤔 

Maybe we need this additional check?
```diff
diff --git a/packages/toolkit/src/query/react/buildHooks.ts b/packages/toolkit/src/query/react/buildHooks.ts
index ba86d843..fbff7d24 100644
--- a/packages/toolkit/src/query/react/buildHooks.ts
+++ b/packages/toolkit/src/query/react/buildHooks.ts
@@ -918,7 +918,12 @@ export function buildHooks<Definitions extends EndpointDefinitions>({
               (_: ApiRootState, lastResult: any) => lastResult,
               (_: ApiRootState) => stableArg,
             ],
-            queryStatePreSelector
+            queryStatePreSelector,
+            {
+              memoizeOptions: {
+                resultEqualityCheck: shallowEqual
+              }
+            }
           ),
         [select, stableArg]
```

**Reply by kkirby** [reply]
I see what you mean now. I wasn't really aware of how `createSelector` worked, but you're right, the expectation is that the function should not be called in this case.

There is a lot going on at play in buildHooks, and I've just started learning about RTK Query, however I _think_ your proposed solution might be correct? The object being passed into `selectFromResult` is different for each render. However, doing a simple stringify of the object shows that the actual data has not changed. So, I think that you're correct: `createSelector` is seeing that the object passed in has changed, even though the properties of it have not. 

-- addendum --

If that is the case (and I'm understanding `createSelector` correctly), you'd probably see a significant speed increase in rendering :P.

**Reply by phryneas** [reply]
This will be fixed with #4029 - thanks to @thisjeremiah for the PR! :)
